src/lib/y2storage/autoinst_proposal.rb

```diff
@@ -0,0 +1,184 @@
+    attr_reader :initial_devicegraph
+
+    # Proposed layout of devices, nil if the proposal has not been
+    # calculated yet
+    # @return [Devicegraph]
+    attr_reader :proposed_devicegraph
+
+    alias_method :devices, :proposed_devicegraph
+
+    # Constructor
+    #
+    # @param partitioning [Array<Hash>] Partitioning schema from an AutoYaST profile
+    # @param devicegraph  [Devicegraph] starting point. If nil, then probed devicegraph
+    #   will be used
+    # @param disk_analyzer [DiskAnalyzer] if nil, a new one will be created
+    #   based in the initial devicegraph
+    def initialize(partitioning: [], devicegraph: nil, disk_analyzer: nil)
+      @partitioning = partitioning
+      @initial_devicegraph = devicegraph
+      @disk_analyzer = disk_analyzer
+      @proposed = false
+    end
+
+    # Checks whether the proposal has already been calculated
+    #
+    # @return [Boolean]
+    def proposed?
+      @proposed
+    end
+
+    # Calculates the proposal
+    #
+    # @raise [UnexpectedCallError] if called more than once
+    # @raise [NoDiskSpaceError] if there is no enough space to perform the installation
+    def propose
+      raise UnexpectedCallError if proposed?
+
+      drives = Proposal::AutoinstDrivesMap.new(initial_devicegraph, partitioning)
+
+      space_maker = Proposal::AutoinstSpaceMaker.new(disk_analyzer)
+      devicegraph = space_maker.provide_space(initial_devicegraph, drives)
+
+      @proposed_devicegraph = propose_devicegraph(devicegraph, drives)
+      @proposed = true
+
+      nil
+    end
+
+  protected
+
+    # Proposes a devicegraph based on given drives map
+    #
+    # This method falls back to #proposed_guided_devicegraph when the device map
+    # does not contain any partition.
+    #
+    # @param devicegraph [Devicegraph]                 Starting point
+    # @param drives      [Proposal::AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Devicegraph containing the planned devices
+    def propose_devicegraph(devicegraph, drives)
+      planned_devices = plan_devices(devicegraph, drives)
+      create_devices(devicegraph, planned_devices, drives.disk_names)
+    rescue Error
+      log.info "No partitions were specified. Falling back to guided setup planning."
+      propose_guided_devicegraph(devicegraph, drives)
+    end
+
+    # Finds a suitable devicegraph using the guided proposal approach
+    #
+    # If the :desired target fails, it will retry with the :min target.
+    #
+    # @param devicegraph [Devicegraph]       Starting point
+    # @param drives      [AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Proposed devicegraph using the guidede proposal approach
+    #
+    # @raise [Error] No suitable devicegraph was found
+    # @see proposed_guided_devicegraph
+    def propose_guided_devicegraph(devicegraph, drives)
+      guided_devicegraph_for_target(devicegraph, drives, :desired)
```

Comment:
Because the planner doesn't know whether is going to be possible to make the devices fit. In the regular (guided) proposal, that's a way more complex process that implies deleting, reusing and shrinking devices between subsequent attempts of making the planned devices fit.

So the planner just set the goal and subsequent steps, in separate classes, try to address it. If it was not possible, a less ambitious goal is set and the process is tried again.

That's the approach followed from the very beginning and [described here](https://github.com/yast/yast-storage-ng/blob/master/doc/proposal.md)... which reminds me another missing task for this pull request - update that document that is not 100% accurate now.